### ISSUE TITLE
/*Validate with renew=true should not succeed when user authenticated via SSO

### ISSUE THREAD

**Issue by bifyu** [apereo/cas#2331]
/*Validate with renew=true should not succeed when user authenticated via SSO
Note: _**renew=true** is required to trigger this behavior._

- CAS server version
  - 5.0.3-SNAPSHOT (AuthN using LDAP, JSON service registry)
- Container/environment type/version (Tomcat, Jetty, Java, Linux/Windows, etc)
  - Tomcat 8.0.39, Java 1.8.0_112
- Steps to duplicate the problem
1. define registered service for testing purposes (JSON)
`{
  "name" : "casRegressionTest",
  "description" : "CAS regression test",
  "serviceId" : "^https://(www\\.)*example\\.com/regression/.*",
  "id" : 10000003,
  "evaluationOrder" : 10,
  "@class" : "org.apereo.cas.services.RegexRegisteredService",}`
3. test service URL 1
`https://example.com/regression/app1`
4. authenticate as user1
5. validate ST for app1
6. without logging out, test service URL 2
`https://example.com/regression/app2`
7. user1authenticated via SSO
8. validate ST for app2 (with renew=true)

CAS validates app2 without requiring authentication (user1 authenticated via SSO) despite renew=true on validation.

Similar results with the /serviceValidate and /samlValidate endpoints

- CAS debug logs
  - [CAS5-sso-renew.log](https://gist.github.com/bifyu/126a90ada2b0dd62fa25d689a7669e1e#file-cas5-sso-renew-log)

**Comment by mmoayyed** [comment]
Can not duplicate.

Here are the steps:

```
- https://sso.example.com/cas/login?service=https://www.app1.com
- Provide credentials. Get an ST-1
- https://sso.example.com/cas/validate?service=https://www.app1.com&ticket=ST-1
```

Validation succeeds. 

Continue with:

```
- https://sso.example.com/cas/login?service=https://www.app2.com
- Skip credentials. Get ST-2
- https://sso.example.com/cas/validate?renew=true&service=https://www.app1.com&ticket=ST-2
```

Validation fails. 

Going to try with 5.0.x too.

**Comment by mmoayyed** [comment]
Same is true with 5.0.x. Validation correctly fails.

**Comment by bifyu** [comment]
Rebuilt again with 5.0.3-SNAPSHOT, I'm still seeing error. In your tests, are you validating for app2? Your comment above says service=app1 for ST-2, but not sure if that's a typo. Here's what I see:

    - https://cas.example.edu/cas/login?service=https://www.example.com/regression/app1
    - provide credentials, get ST-1
    - https://cas-dev.its.hawaii.edu/cas/validate?service=https://www.example.com/regression/app1&ticket=ST-1

Validation succeeds.
Continue with:

    - https://cas-dev.its.hawaii.edu/cas/login?service=https://www.example.com/regression/app2
    - SSO, skip credentials, get ST-2
    - https://cas.example.edu/cas/validate?service=https://www.example.com/regression/app2&ticket=ST-2&renew=true

Validation succeeds.

I notice one other difference is that we have renew=true at the end of the parameters, whereas you had it first. Is it possible the order matters?

**Comment by mmoayyed** [comment]
>  In your tests, are you validating for app2? Your comment above says service=app1 for validation of ST-2, but not sure if that's a typo

I certainly did validate. The comment was a typo, but I will run through this tomorrow again with more care to make sure I didn't miss anything.

>  Is it possible the order matters?

Does not.

> This seems to be new, don't recall seeing this previously

Yes the log statement is new; added to really see what's being pass to the spec. It's saying that `Cas10ProtocolValidationSpecification` is asked to enforce `renew` protocol bebavior by making sure the assertion is generated from a new authentication session with user putting in credentials explicitly. What you described basically in simpler terms :) 

Logs do indicate buggy behavior for sure. I'll see if I can reproduce this.

**Comment by tsschmidt** [comment]
I saw this behavior in both 4.2.6 build and 5.1.0-SNAPSHOT-RC2 from master.  Here are the logs:

2017-01-27 15:47:41,279 DEBUG [org.apereo.cas.validation.Cas20WithoutProxyingValidationSpecification] - <Is validation specification set to enforce [renew] protocol behavior? [yes]. Is assertion issued from a new login? [yes]>
2017-01-27 15:47:41,279 DEBUG [org.apereo.cas.validation.Cas20WithoutProxyingValidationSpecification] - <Number of chained authentications in the assertion [1]>
2017-01-27 15:47:41,279 DEBUG [org.apereo.cas.validation.Cas20WithoutProxyingValidationSpecification] - <Validation specification is satisfied by the produced assertion>

fromNewLogin was true, but I did not enter credentials.  Opened a new tab and hit a new service.  Possible problem with fromNewLogin getting set our passed?

**Comment by mmoayyed** [comment]
This is really a regression caused by the fix applied here: https://github.com/apereo/cas/issues/2321 I have fixed this issue, but now will need to circle back to https://github.com/apereo/cas/issues/2321 and find a better solution.

**Comment by mmoayyed** [comment]
You should be able to test the fix in about an hour or so, by using the new SNAPSHOT. I tested both scenarios reported at #2321 and #2331. All checks out. I'll begin to port the patch back once the fixes can be confirmed in master. @tsschmidt could you try again please?

**Comment by bifyu** [comment]
Rebuilt again with 5.0.3-SNAPSHOT, I'm still seeing error?

    2017-01-30 14:26:43,841 DEBUG [org.apereo.cas.services.RegisteredServiceAccessStrategyUtils] - <Current authentication via ticket TGT-**********************************************ZmS2gesq9p-cas allows service https://www.example.com/regression/app2 to participate in the existing SSO session>
    2017-01-30 14:26:43,843 INFO [org.apereo.cas.CentralAuthenticationServiceImpl] - <Granted ticket [ST-2-KcVCXI0lBABjxxKr20d0-cas] for service [https://www.example.com/regression/app2] and principal [user1]>
    2017-01-30 14:26:43,915 DEBUG [org.apereo.cas.validation.Cas10ProtocolValidationSpecification] - <Is validation specification set to enforce [renew] protocol behavior? [yes]. Is assertion issued from a new login? [yes]>
    2017-01-30 14:26:43,916 DEBUG [org.apereo.cas.web.LegacyValidateController] - <Successfully validated service ticket ST-2-KcVCXI0lBABjxxKr20d0-cas for service [https://www.example.com/regression/app2]>

**Comment by mmoayyed** [comment]
Sorry, by SNAPSHOT I meant the current master. Not backported yet; pending @tsschmidt or someone else's confirmation.

**Comment by tsschmidt** [comment]
Looks to be working.  I get Ticket not recognized message when adding 
renew=true on the serviceValidate when not from new login.

**Comment by mmoayyed** [comment]
Thanks. I'll try to backport later today. Will post back when done.

**Comment by mmoayyed** [comment]
Starting to backport patch to 5.0.x. Should be good to go by tomorrow

**Comment by mmoayyed** [comment]
Fix backported.

**Comment by bifyu** [comment]
Appears fixed
